### DISCUSSION TITLE
Inconsistent Preloading Behavior with User Script

### DISCUSSION THREAD

**Discussion by jameskuszmaul-brt** [foxglove/community#855] [Q&A] [answered]
I a currently able to sometimes trigger weird behaviors around proper preloading behavior for plot panels. With the below layout & MCAP file, if I:

1. Go to studio.foxglove.dev
2. Open the MCAP file.
3. Load the layout; it generally seems to load correctly on the first try.
4. Select on the bottom slider a timepoint somewhere in the middle of the log (this does seem to be important).
5. Swap to the second tab & then swap back.
6. I then will generally observe that either the user script output (the second plot) or the raw data itself (the top plot) doesn't load fully. In general if I seek the playback somewhere and press "play" the plot will start populating correctly.

For reference, all the layout consists of is:
* Two tabs (so I can easily make things load/unload; although I too would love to see https://github.com/orgs/foxglove/discussions/846 resolved).
* One tab has a plot panel with some data from the MCAP file (the left/right voltage fields); a plot with the output of the user script; and a state transition panel with some other topic (I have no idea if this is important, but it's nice as a reference for what is working).
* The user script takes as an input topic the data that is in the first plot; it performs a very simple transformation (just divides some numbers by 12), and then publishes to the topic that is used in the second plot.

Failure mode where the user script data still loaded correctly, but the only loaded data directly from the MCAP file is from when I had seeked in the file and hit play for a bit:
![image](https://github.com/foxglove/community/assets/91565539/af4a3f10-e919-44c6-8a6a-8be78cb58326)
Failure mode where only parts of each plot had loaded, for some reason:
![image](https://github.com/foxglove/community/assets/91565539/0dfe6119-6b07-4778-8f0d-9c7d372cb7ed)

User script:
[Bad Plot_User Script Loading.json](https://github.com/foxglove/community/files/13044423/Bad.Plot_User.Script.Loading.json)

MCAP file: ~30M XZ compressed:
https://drive.google.com/file/d/1a5hTuSFBAJl5pgK3-fR0l2MfQyUu8Nf6/view?usp=sharing

**Comment by cfoust** [comment]
Looking into this right now.

**Comment by cfoust** [comment, marked as answer]
This turned out to be rather complicated, but I'm pretty close to finishing up [the fix](https://github.com/foxglove/studio/pull/7041).
